### PR TITLE
[jit] fix handling of function attributes.

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28395 [jit] fix @property
* **#28569 [jit] fix handling of function attributes.**

Previously, the inclusion of function attributes would "poison" a
ConcreteModuleType, because we did not have a way of checking whether
they are actually the same function. This PR uses the Python function
object to perform that check. This improves our ability to reuse JIT
types between modules.

Also this PR fixes a bug where we weren't properly adding modules as
attributes when converting from ConcreteType -> JIT type (we were adding
them after the fact--another reason to switch from using `register_x` to
`set_x` during module construction, which is on my to-do list after
this).

Fixes https://github.com/pytorch/pytorch/issues/28559

Pull Request resolved: https://github.com/pytorch/pytorch/pull/28569

Differential Revision: [D18111331](https://our.internmc.facebook.com/intern/diff/D18111331)

### CODE REVIEW

File: test/jit/test_type_sharing.py [hunk 1]
```diff
@@ -301,6 +322,26 @@ def forward(self, x):
 
         self.assertDifferentType(fn1_mod, fn2_mod)
 
+    def test_python_function_attribute_same(self):
+        """
+        Different functions passed in should lead to different types
```
Comment (driazati) [inline test/jit/test_type_sharing.py]: Different -> same?

File: torch/csrc/jit/script/concrete_module_type.cpp [hunk 2]
```diff
@@ -36,6 +36,11 @@ ClassTypePtr ConcreteModuleType::createNewTypeFromThis() {
     cls->addAttribute(name, type, isParameter);
   }
 
+  for (const auto& moduleInfo : modules_) {
```
Comment (driazati) [inline torch/csrc/jit/script/concrete_module_type.cpp]: What is this adding to `cls`? Is `modules_` all callables attached to this type or is it just sub-`ScriptModule`s?

**Review by driazati — APPROVED** [review]

File: test/jit/test_type_sharing.py [hunk 3]
```diff
@@ -301,6 +322,26 @@ def forward(self, x):
...
+                super(M, self).__init__()
+                self.fn = fn
+
+            def forward(self, x):
+                return self.fn(x)
+
+        fn1_mod = M(fn)
+        fn2_mod = M(fn)
+
+        self.assertSameType(fn1_mod, fn2_mod)
```
[the remark this answers is not in the document]
Comment (suo) [reply test/jit/test_type_sharing.py]: that's test_python_function_attribute_different

File: torch/csrc/jit/script/concrete_module_type.cpp (hunk 2 above)
Comment (suo) [reply torch/csrc/jit/script/concrete_module_type.cpp]: it's all the submodules

File: test/jit/test_recursive_script.py [hunk 4]
```diff
@@ -531,6 +531,30 @@ def forward(self, x):
...
+            def forward(self, src):
+                output = src
+                output = self.norm(output)
+                return output
+
+        class M(nn.Module):
+            def __init__(self):
+                super().__init__()
+                encoder_norm = nn.ReLU()
+                self.encoder = N(encoder_norm)
```
[the remark this answers is not in the document]
Comment (suo) [reply test/jit/test_recursive_script.py]: oh my attempt to minify this test case went wrong. I'll add back the function attr

**Pull request closed without merging**